### PR TITLE
Refactor TreeView

### PR DESCRIPTION
**Pull request by StanleyGoldman**
**Note**: This branch targets `enhancements/changes-tree-view-rollup` #471 

Initially tree loading functionality was isolated to `GitHub.Api` in order to be able to be tested.
This is not the only functionality around `Tree` that needs to be testable.
The functionality to toggle visibility and checked state is also quite complex and could serve to be isolated.
Also with some crafty use of generics we can make the additional tree types easier to implement.

- Created `TreeBase<TNode, TData>`
  - Changed `Tree` to `Tree<TNode, TData>: TreeBase<TNode, TData>`
  - Moved `Load` from `TreeLoader`
    - Deleted `ITree`
  - Moved `ToggleNodeVisibility`, `ToggleNodeChecked` from `Tree` to `TreeBase`

Additionally added an `IconBadge` to `TreeNode` and it's `Render`

### CODE REVIEW

File: src/GitHub.Api/UI/TreeBase.cs [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+            var nodeLevel = node.Level;
+            node.IsCollapsed = !node.IsCollapsed;
+            idx++;
+            for (; idx < Nodes.Count && Nodes[idx].Level > nodeLevel; idx++)
+            {
+                Nodes[idx].IsHidden = node.IsCollapsed;
+                if (Nodes[idx].IsFolder && !node.IsCollapsed && Nodes[idx].IsCollapsed)
+                {
+                    var level = Nodes[idx].Level;
+                    for (idx++; idx < Nodes.Count && Nodes[idx].Level > level; idx++)
```
Comment (sguthals) [inline src/GitHub.Api/UI/TreeBase.cs]: Is this just to get to the last node?

File: src/GitHub.Api/UI/TreeBase.cs [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+            switch (node.CheckState)
+            {
+                case CheckState.Mixed:
+                case CheckState.Empty:
+                    node.CheckState = CheckState.Checked;
+                    isChecked = true;
+                    break;
+
+                case CheckState.Checked:
```
Comment (sguthals) [inline src/GitHub.Api/UI/TreeBase.cs]: You don't need to `isChecked = false;` here?

File: src/GitHub.Api/UI/TreeBase.cs (hunk 1 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/UI/TreeBase.cs]: Hmm. This is @shana's logic which i'm moving from place to place.

File: src/GitHub.Api/UI/TreeBase.cs (hunk 2 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/UI/TreeBase.cs]: It is set to false by default a few lines up.

File: src/GitHub.Api/UI/TreeBase.cs (hunk 1 above)
Comment (StanleyGoldman) [reply src/GitHub.Api/UI/TreeBase.cs]: lol... uhmm yeaa... she is going to have to document that...

Comment (StanleyGoldman) [reply src/GitHub.Api/UI/TreeBase.cs]: We can add tests for it now.. 🐑

File: src/GitHub.Api/UI/TreeBase.cs (hunk 2 above)
Comment (sguthals) [reply src/GitHub.Api/UI/TreeBase.cs]: Duh..

**Pull request merged**